nixos/modules/system/boot/systemd-nspawn.nix

```diff
@@ -105,29 +226,46 @@ in {
+      description = ''
+        Definition of <citerefentry><refentrytitle>systemd.nspawn</refentrytitle>
+        <manvolnum>5</manvolnum></citerefentry> configurations.
+      '';
+    };
+
+    systemd.disableUnprivilegedNspawn = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        If true, it disables unprivileged systemd-nspawn container so that
+        <literal>/nix</literal> mountpoint in guest NixOS works correctly.
+        It removes <literal>-U</literal> or
+        <literal>--private-users=pick --private-users-chown</literal> options
+        in <literal>serviceConfig.ExecStart</literal> by overriding upstream
+        <literal>systemd-nspawn@.service</literal>.
+        Default is false, because it can break <literal>systemd-nspawn@*.service</literal> having
+        non-default <literal>serviceConfig.ExecStart</literal>.
+      '';
     };
 
   };
 
   config =
-    let
-      units = mapAttrs' (n: v: let nspawnFile = "${n}.nspawn"; in nameValuePair nspawnFile (instanceToUnit nspawnFile v)) cfg;
-    in 
-      mkMerge [
-        (mkIf (cfg != {}) { 
-          environment.etc."systemd/nspawn".source = mkIf (cfg != {}) (generateUnits' false "nspawn" units [] []);
-        })
-        {
-          systemd.targets.multi-user.wants = [ "machines.target" ];
-
-          # Workaround for https://github.com/NixOS/nixpkgs/pull/67232#issuecomment-531315437 and https://github.com/systemd/systemd/issues/13622
-          # Once systemd fixes this upstream, we can re-enable -U
-          systemd.services."systemd-nspawn@".serviceConfig.ExecStart = [ 
-            ""  # deliberately empty. signals systemd to override the ExecStart
-            # Only difference between upstream is that we do not pass the -U flag
-            "${config.systemd.package}/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --network-veth --settings=override --machine=%i"
-          ];
-        }
-      ];
+    mkMerge [
+      (mkIf (cfg != {}) {
+        environment.etc = mapAttrs' (n: v:
+          let nspawnFile = "${n}.nspawn";
+          in nameValuePair "systemd/nspawn/${nspawnFile}" (instanceToUnit nspawnFile v)
+          ) cfg;
+      })
+      { systemd.targets.multi-user.wants = [ "machines.target" ]; }
+      (mkIf config.systemd.disableUnprivilegedNspawn {
+        # Workaround for https://github.com/NixOS/nixpkgs/pull/67232#issuecomment-531315437 and https://github.com/systemd/systemd/issues/13622
+        # Once systemd fixes this upstream, we can re-enable -U
+        systemd.services."systemd-nspawn@".serviceConfig.ExecStart = [
+          ""  # deliberately empty. signals systemd to override the ExecStart
+          # Only difference between upstream is that we do not pass the -U flag
+          "${config.systemd.package}/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --network-veth --settings=override --machine=%i"
```

Comment:
@flokli do you know if the systemd bug which made this necessary is fixed? Also, I'd prefer to set `disableUnprivilegedNspawn` to `true` by default since it's been a valid workaround for us.